mauth_client/middlewares/asgi.py

```diff
@@ -0,0 +1,91 @@
+import json
+import logging
+
+from asgiref.typing import (
+    ASGI3Application,
+    ASGIReceiveCallable,
+    ASGISendCallable,
+    Scope,
+)
+
+from mauth_client.authenticator import LocalAuthenticator
+from mauth_client.config import Config
+from mauth_client.consts import (
+    ENV_APP_UUID,
+    ENV_AUTHENTIC,
+    ENV_PROTOCOL_VERSION,
+)
+from mauth_client.signable import RequestSignable
+from mauth_client.signed import Signed
+from mauth_client.utils import decode
+
+logger = logging.getLogger("mauth_asgi")
+
+
+class MAuthASGIMiddleware:
+    def __init__(self, app: ASGI3Application, exempt: set = set()) -> None:
+        self._validate_configs()
+        self.app = app
+        self.exempt = exempt
+
+    async def __call__(
+        self, scope: Scope, receive: ASGIReceiveCallable, send: ASGISendCallable
+    ) -> None:
+        path = scope["path"]
+
+        if path in self.exempt:
+            return await self.app(scope, receive, send)
+
+        query_string = scope["query_string"]
+        url = f"{path}?{decode(query_string)}" if query_string else path
+        headers = {decode(k): decode(v) for k, v in scope["headers"]}
+        body = await self._get_body(receive)
+
+        signable = RequestSignable(
+            method=scope["method"],
+            url=url,
+            body=body,
+        )
+        signed = Signed.from_headers(headers)
+        authenticator = LocalAuthenticator(signable, signed, logger)
+        is_authentic, status, message = authenticator.is_authentic()
+
+        if is_authentic:
+            # asgi spec calls for passing a copy of the scope rather than mutating it
+            # note: deepcopy will blow up with infi recursion due to objects in some values
+            scope_copy = scope.copy()
+            scope_copy[ENV_APP_UUID] = signed.app_uuid
+            scope_copy[ENV_AUTHENTIC] = True
+            scope_copy[ENV_PROTOCOL_VERSION] = signed.protocol_version()
+            await self.app(scope_copy, receive, send)
+        else:
+            await self._send_response(send, status, message)
+
+    def _validate_configs(self) -> None:
+        # Validate the client settings (APP_UUID, PRIVATE_KEY)
+        if not all([Config.APP_UUID, Config.PRIVATE_KEY]):
+            raise TypeError("MAuthASGIMiddleware requires APP_UUID and PRIVATE_KEY")
+        # Validate the mauth settings (MAUTH_BASE_URL, MAUTH_API_VERSION)
+        if not all([Config.MAUTH_URL, Config.MAUTH_API_VERSION]):
+            raise TypeError("MAuthASGIMiddleware requires MAUTH_URL and MAUTH_API_VERSION")
+
+    async def _get_body(self, receive: ASGIReceiveCallable) -> str:
+        body = b""
+        more_body = True
+        while more_body:
+            msg = await receive()
+            body += msg.get("body", b"")
+            more_body = msg.get("more_body", False)
+        return decode(body)
```

Comment:
Won't this throw an exception if a middleware higher up the stack has already read in the body?

Reply:
ha, it will actually hang forever `await`ing the body since there is nothing left on the stream.  thanks for pointing this out, gonna have to go spelunking in the source for starlette/fastapi.  we only have the `receive` function that lets us consume events from the stream.  my guess is that there is some "by convention" location on the `scope` where we have to stick the body when reading it (and look for before reading in case some prior middleware already read).

Reply:
I think we don't need to decode `body` here

Reply:
with further thought it seems like the correct approach is to read the body into memory and then pass a fake `receive` func to downstream stuff that will read from our array of events in memory rather than the actual stream..  this will ofc defeat one of the advantages of being able to process req bodies in a streaming fashion in frameworks that use this middleware (since our middleware will have to consume it all to check the auth before passing control).

Reply:
@ykitamura-mdsol so the authenticator can handle `bytes`?  Does that mean we also don't need to decode the headers or query string?

Reply:
just `body`. probably this PR explains it #7